feeds/wlan-ap/apc/src/src/ubus.c

```diff
@@ -95,38 +99,67 @@ apc_info_handle(struct ubus_context *ctx, struct ubus_object *obj,
 }
 
 static char apc_mode[APC_MAX_MODE][8] = {"DOWN", "LOOP", "WT", "PTP", "OR", "BDR", "DR"};
-void apc_update_state()
+int apc_update_state(void)
 {
 	struct in_addr dr_addr;
 	struct in_addr bdr_addr;
-	dr_addr.s_addr = htonl(apc_ifa->drip);
-	bdr_addr.s_addr = htonl(apc_ifa->bdrip);
+	ip_addr cur_drip;
+	ip_addr cur_bdrip;
+	u8 cur_state;
+
+	cur_drip = apc_ifa->drip;
+	cur_bdrip = apc_ifa->bdrip;
+	cur_state = apc_ifa->state;
+	ucount++;
+
+	if (cur_drip == old_drip &&
+	    cur_bdrip == old_bdrip &&
+	    cur_state == old_state && ucount < APC_NOTIFY_INTERVAL) {
+		printf("NOT changed");
+		return -1;
+	}
+
+	printf("Changed %u", ucount);
+	ucount = 0;
+
+	dr_addr.s_addr = htonl(cur_drip);
+	bdr_addr.s_addr = htonl(cur_bdrip);
 
 	state.enabled = true;
 	if ((apc_ifa->state == APC_IS_DR) ||
 	    (apc_ifa->state == APC_IS_BACKUP) ||
 	    (apc_ifa->state == APC_IS_DROTHER)) {
 		snprintf(state.mode, sizeof(state.mode), "%s",
-			 &apc_mode[apc_ifa->state][0]);
+			 &apc_mode[cur_state][0]);
 		snprintf(state.dr_addr, sizeof(state.dr_addr),
 			 "%s", inet_ntoa(dr_addr));
 		snprintf(state.bdr_addr, sizeof(state.bdr_addr),
 			 "%s", inet_ntoa(bdr_addr));
 	} else if (apc_ifa->state == APC_IS_WAITING) {
 		snprintf(state.mode, sizeof(state.mode), "%s",
-			 &apc_mode[apc_ifa->state][0]);
+			 &apc_mode[cur_state][0]);
 		snprintf(state.dr_addr, sizeof(state.dr_addr), "0.0.0.0");
 		snprintf(state.bdr_addr, sizeof(state.bdr_addr), "0.0.0.0");
 	} else {
 		snprintf(state.mode, sizeof(state.mode), "NC");
 		snprintf(state.dr_addr, sizeof(state.dr_addr), "0.0.0.0");
 		snprintf(state.bdr_addr, sizeof(state.bdr_addr), "0.0.0.0");
 	}
+
+	old_drip = cur_drip;
+	old_bdrip = cur_bdrip;
+	old_state = cur_state;
```

Comment:
Did anything change the underlying apc_ifa->drip, apc_ifa->bdrip, apc_ifa->state variables?
Does 'state' variable point at part of apc_ifa or something?
Otherwise, I don't think the cur_drip, cur_bdrip and cur_state variables are actually required.

Reply:
The apc_ifa->state, apc_ifa->drip,apc_ifa->bdrip might change while this code is executing. So I want to work with the same copy of these variables throughout this function.